#  Filtering Operations

**Learning Objectives:**
  * Gain an introduction to filtering operation using the *pandas* library
  * Slice, dice and summarize data within a `DataFrame` applying basic filters
  

## Library import

The following line imports the *pandas* library

In [ ]:
import pandas as pd


## Data loading and DataFrame creation


. The following example loads a file with the famous titanic data. Run the following cell to load the data and create your first `DataFrame`

In [ ]:
titanic = pd.read_csv("https://raw.githubusercontent.com/thousandoaks/Python4DS101/master/data/titanic.csv", sep=",")


Let's display the first few records of a `DataFrame`:

In [ ]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

The data consists of the following data columns:

* PassengerId: Id of every passenger.

* Survived: This feature have value 0 and 1. 0 for not survived and 1 for survived.

* Pclass: There are 3 classes: Class 1, Class 2 and Class 3.

* Name: Name of passenger.

* Sex: Gender of passenger.

* Age: Age of passenger.

* SibSp: Indication that passenger have siblings and spouse.

* Parch: Whether a passenger is alone or have family.

* Ticket: Ticket number of passenger.

* Fare: Indicating the fare.

* Cabin: The cabin of passenger.

* Embarked: The embarked category.





## Filtering Operations

Composite filtering operations in `Pandas` are implemented in three stages:
* Filter definition
* Filter saving
* Filter application



<img src='https://pandas.pydata.org/docs/_images/03_subset_rows.svg'>

For instance in the following example we select passengers older than 35 years



### Stage 1. We define the filter set
#### We use the boolean operator `larger than` to select passenger whose Age column is larger than 35
#### We use the boolean operator `equal to` to select passengers that survived

In [ ]:
titanic["Age"] > 35

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

### Stage 2. We save the filters

In [ ]:
AgeFilter=titanic["Age"] > 35

### Stage 3. We apply the filter

In [ ]:
titanic[AgeFilter]

PassengerId  Survived  Pclass  \
1              2         1       1   
6              7         0       1   
11            12         1       1   
13            14         0       3   
15            16         1       2   
..           ...       ...     ...   
865          866         1       2   
871          872         1       1   
873          874         0       3   
879          880         1       1   
885          886         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
..                                                 ...     ...   ...    ...   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
6        0     17463  51.8625   E46        S  
11       0    113783  26.5500  C103        S  
13       5    347082  31.2750   NaN        S  
15       0    248706  16.0000   NaN        S  
..     ...       ...      ...   ...      ...  
865      0    236852  13.0000   NaN        S  
871      1     11751  52.5542   D35        S  
873      0    345765   9.0000   NaN        S  
879      1     11767  83.1583   C50        C  
885      5    382652  29.1250   NaN        Q  

[217 rows x 12 columns]

The filter returned 217 passengers older than 35 who survived